shadow/man/man1/chfn.1
'\" t
.\"     Title: chfn
.\"    Author: Julianne Frances Haugh
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 01/19/2026
.\"    Manual: User Commands
.\"    Source: shadow-utils 4.19.2
.\"  Language: English
.\"
.TH "CHFN" "1" "01/19/2026" "shadow\-utils 4\&.19\&.2" "User Commands"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
chfn \- change full user name and information
.SH "SYNOPSIS"
.HP \w'\fBchfn\fR\ 'u
\fBchfn\fR [\fIoptions\fR] [\fILOGIN\fR]
.SH "DESCRIPTION"
.PP
The
\fBchfn\fR
command changes the user\*(Aqs full name, office room number, office phone number, and home phone number information for an account in the respective fields of
/etc/passwd\&. This information is typically printed by
\fBfinger\fR(1)
and similar programs\&. A normal user may only change the fields for her own account, subject to the restrictions in
/etc/login\&.defs\&. (The default configuration is to prevent users from changing their fullname\&.) The superuser may change any field for any account\&. Additionally, only the superuser may use the
\fB\-o\fR
option to change the undefined portions of the GECOS field\&.
.PP
These fields must not contain any colons\&. Except for the
\fIother\fR
field, they should not contain any comma or equal sign\&. It is also recommended to avoid non\-US\-ASCII characters, but this is only enforced for the phone numbers\&. The
\fIother\fR
field is used to store accounting information used by other applications\&.
.PP
If
\fILOGIN\fR
is not specified,
\fBchfn\fR
acts on the current user, and prompts for the current user\*(Aqs password\&.
.SH "OPTIONS"
.PP
The options which apply to the
\fBchfn\fR
command are:
.PP
\fB\-f\fR, \fB\-\-full\-name\fR\ \&\fIFULL_NAME\fR
.RS 4
Change the user\*(Aqs full name\&.
.RE
.PP
\fB\-h\fR, \fB\-\-home\-phone\fR\ \&\fIHOME_PHONE\fR
.RS 4
Change the user\*(Aqs home phone number\&.
.RE
.PP
\fB\-o\fR, \fB\-\-other\fR\ \&\fIOTHER\fR
.RS 4
Change the user\*(Aqs other GECOS information\&. This field is used to store accounting information used by other applications, and can be changed only by a superuser\&.
.RE
.PP
\fB\-r\fR, \fB\-\-room\fR\ \&\fIROOM_NUMBER\fR
.RS 4
Change the user\*(Aqs room number\&.
.RE
.PP
\fB\-R\fR, \fB\-\-root\fR\ \&\fICHROOT_DIR\fR
.RS 4
Apply changes in the
\fICHROOT_DIR\fR
directory and use the configuration files from the
\fICHROOT_DIR\fR
directory\&. No SELINUX support\&.
.RE
.PP
\fB\-u\fR, \fB\-\-help\fR
.RS 4
Display help message and exit\&.
.RE
.PP
\fB\-w\fR, \fB\-\-work\-phone\fR\ \&\fIWORK_PHONE\fR
.RS 4
Change the user\*(Aqs office phone number\&.
.RE
.PP
If none of the options is selected,
\fBchfn\fR
operates in an interactive fashion, prompting the user with the current values for all of the fields\&. Enter the new value to change the field, or leave the line blank to use the current value\&. The current value is displayed between a pair of
\fB []\fR
marks\&.
.SH "CONFIGURATION"
.PP
The following configuration variables in
/etc/login\&.defs
change the behavior of this tool:
.PP
\fBCHFN_AUTH\fR (boolean)
.RS 4
If
\fIyes\fR, the
\fBchfn\fR
program will require authentication before making any changes, unless run by the superuser\&.
.RE
.PP
\fBCHFN_RESTRICT\fR (string)
.RS 4
This parameter specifies which values in the
\fIgecos\fR
field of the
/etc/passwd
file may be changed by regular users using the
\fBchfn\fR
program\&. It can be any combination of letters
\fIf\fR,
\fIr\fR,
\fIw\fR,
\fIh\fR, for Full name, Room number, Work phone, and Home phone, respectively\&. For backward compatibility,
\fIyes\fR
is equivalent to
\fIrwh\fR
and
\fIno\fR
is equivalent to
\fIfrwh\fR\&. If not specified, only the superuser can make any changes\&. The most restrictive setting is better achieved by not installing
\fBchfn\fR
SUID\&.
.RE
.PP
\fBLOGIN_STRING\fR (string)
.RS 4
The string used for prompting a password\&. The default is to use "Password: ", or a translation of that string\&. If you set this variable, the prompt will not be translated\&.
.sp
If the string contains
\fI%s\fR, this will be replaced by the user\*(Aqs name\&.
.RE
.SH "FILES"
.PP
/etc/login\&.defs
.RS 4
Shadow password suite configuration\&.
.RE
.PP
/etc/passwd
.RS 4
User account information\&.
.RE
.SH "SEE ALSO"
.PP
\fBchsh\fR(1),
\fBlogin.defs\fR(5),
\fBpasswd\fR(5)\&.


shadow/src/chfn.c
/*
 * SPDX-FileCopyrightText: 1989 - 1994, Julianne Frances Haugh
 * SPDX-FileCopyrightText: 1996 - 2000, Marek Michałkiewicz
 * SPDX-FileCopyrightText: 2001 - 2006, Tomasz Kłoczko
 * SPDX-FileCopyrightText: 2007 - 2011, Nicolas François
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include "config.h"

#ident "$Id$"

#include <fcntl.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <getopt.h>

#include "chkname.h"
#include "defines.h"
/*@-exitarg@*/
#include "exitcodes.h"
#include "fields.h"
#include "getdef.h"
#include "nscd.h"
#ifdef USE_PAM
#include "pam_defs.h"
#endif
#include "prototypes.h"
#include "pwauth.h"
#include "pwio.h"
#include "shadowlog.h"
#include "sizeof.h"
#include "sssd.h"
#include "string/sprintf/stpeprintf.h"
#include "string/strcmp/streq.h"
#include "string/strcpy/strtcpy.h"
#include "string/strdup/strdup.h"

struct option_flags {
	bool chroot;
};

/*
 * Global variables.
 */
static const char Prog[] = "chfn";
static char fullnm[BUFSIZ];
static char roomno[BUFSIZ];
static char workph[BUFSIZ];
static char homeph[BUFSIZ];
static char slop[BUFSIZ];
static bool amroot;
/* Flags */
static bool fflg = false;		/* -f - set full name                */
static bool rflg = false;		/* -r - set room number              */
static bool wflg = false;		/* -w - set work phone number        */
static bool hflg = false;		/* -h - set home phone number        */
static bool oflg = false;		/* -o - set other information        */
static bool pw_locked = false;

/*
 * External identifiers
 */

/* local function prototypes */
NORETURN static void fail_exit (int code, bool process_selinux);
NORETURN static void usage (int status);
static bool may_change_field (int);
static void new_fields (void);
static void process_flags (int argc, char **argv, struct option_flags *flags);
static void check_perms (const struct passwd *pw);
static void update_gecos(const char *user, char *gecos, const struct option_flags *flags);
static void get_old_fields (const char *gecos);

/*
 * fail_exit - exit with an error and do some cleanup
 */
static void fail_exit (int code, bool process_selinux)
{
	if (pw_locked) {
		if (pw_unlock (process_selinux) == 0) {
			fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, pw_dbname ());
			SYSLOG ((LOG_ERR, "failed to unlock %s", pw_dbname ()));
			/* continue */
		}
	}
	pw_locked = false;

	closelog ();

	exit (code);
}

/*
 * usage - print command line syntax and exit
 */
NORETURN
static void
usage (int status)
{
	FILE *usageout = (E_SUCCESS != status) ? stderr : stdout;
	(void) fprintf (usageout,
	                _("Usage: %s [options] [LOGIN]\n"
	                  "\n"
	                  "Options:\n"),
	                Prog);
	(void) fputs (_("  -f, --full-name FULL_NAME     change user's full name\n"), usageout);
	(void) fputs (_("  -h, --home-phone HOME_PHONE   change user's home phone number\n"), usageout);
	(void) fputs (_("  -o, --other OTHER_INFO        change user's other GECOS information\n"), usageout);
	(void) fputs (_("  -r, --room ROOM_NUMBER        change user's room number\n"), usageout);
	(void) fputs (_("  -R, --root CHROOT_DIR         directory to chroot into\n"), usageout);
	(void) fputs (_("  -u, --help                    display this help message and exit\n"), usageout);
	(void) fputs (_("  -w, --work-phone WORK_PHONE   change user's office phone number\n"), usageout);
	(void) fputs ("\n", usageout);
	exit (status);
}

/*
 * may_change_field - indicate if the user is allowed to change a given field
 *                    of her gecos information
 *
 *	root can change any field.
 *
 *	field should be one of 'f', 'r', 'w', 'h'
 *
 *	Return true if the user can change the field and false otherwise.
 */
static bool may_change_field (int field)
{
	const char *cp;

	/*
	 * CHFN_RESTRICT can now specify exactly which fields may be changed
	 * by regular users, by using any combination of the following
	 * letters:
	 *  f - full name
	 *  r - room number
	 *  w - work phone
	 *  h - home phone
	 *
	 * This makes it possible to disallow changing the room number
	 * information, for example - this feature was suggested by Maciej
	 * 'Tycoon' Majchrowski.
	 *
	 * For backward compatibility, "yes" is equivalent to "rwh",
	 * "no" is equivalent to "frwh". Only root can change anything
	 * if the string is empty or not defined at all.
	 */
	if (amroot) {
		return true;
	}

	cp = getdef_str ("CHFN_RESTRICT");
	if (NULL == cp) {
		cp = "";
	} else if (streq(cp, "yes")) {
		cp = "rwh";
	} else if (streq(cp, "no")) {
		cp = "frwh";
	}

	if (strchr(cp, field))
		return true;

	return false;
}

/*
 * new_fields - change the user's GECOS information interactively
 *
 * prompt the user for each of the four fields and fill in the fields from
 * the user's response, or leave alone if nothing was entered.
 */
static void new_fields (void)
{
	puts (_("Enter the new value, or press ENTER for the default"));

	if (may_change_field ('f')) {
		change_field(fullnm, sizeof(fullnm), _("Full Name"));
	} else {
		printf (_("\t%s: %s\n"), _("Full Name"), fullnm);
	}

	if (may_change_field ('r')) {
		change_field(roomno, sizeof(roomno), _("Room Number"));
	} else {
		printf (_("\t%s: %s\n"), _("Room Number"), roomno);
	}

	if (may_change_field ('w')) {
		change_field(workph, sizeof(workph), _("Work Phone"));
	} else {
		printf (_("\t%s: %s\n"), _("Work Phone"), workph);
	}

	if (may_change_field ('h')) {
		change_field(homeph, sizeof(homeph), _("Home Phone"));
	} else {
		printf (_("\t%s: %s\n"), _("Home Phone"), homeph);
	}

	if (amroot) {
		change_field(slop, sizeof(slop), _("Other"));
	}
}

/*
 * process_flags - parse the command line options
 *
 *	It will not return if an error is encountered.
 */
static void process_flags (int argc, char **argv, struct option_flags *flags)
{
	int c;		/* flag currently being processed    */
	static struct option long_options[] = {
		{"full-name",  required_argument, NULL, 'f'},
		{"home-phone", required_argument, NULL, 'h'},
		{"other",      required_argument, NULL, 'o'},
		{"room",       required_argument, NULL, 'r'},
		{"root",       required_argument, NULL, 'R'},
		{"help",       no_argument,       NULL, 'u'},
		{"work-phone", required_argument, NULL, 'w'},
		{NULL, 0, NULL, '\0'}
	};

	/*
	 * The remaining arguments will be processed one by one and executed
	 * by this command. The name is the last argument if it does not
	 * begin with a "-", otherwise the name is determined from the
	 * environment and must agree with the real UID. Also, the UID will
	 * be checked for any commands which are restricted to root only.
	 */
	while ((c = getopt_long (argc, argv, "f:h:o:r:R:uw:",
	                         long_options, NULL)) != -1) {
		switch (c) {
		case 'f':
			if (!may_change_field ('f')) {
				fprintf (stderr,
				         _("%s: Permission denied.\n"), Prog);
				exit (E_NOPERM);
			}
			fflg = true;
			strtcpy_a(fullnm, optarg);
			break;
		case 'h':
			if (!may_change_field ('h')) {
				fprintf (stderr,
				         _("%s: Permission denied.\n"), Prog);
				exit (E_NOPERM);
			}
			hflg = true;
			strtcpy_a(homeph, optarg);
			break;
		case 'o':
			if (!amroot) {
				fprintf (stderr,
				         _("%s: Permission denied.\n"), Prog);
				exit (E_NOPERM);
			}
			oflg = true;
			if (strlen (optarg) > (unsigned int) 80) {
				fprintf (stderr,
				         _("%s: fields too long\n"), Prog);
				exit (E_NOPERM);
			}
			strtcpy_a(slop, optarg);
			break;
		case 'r':
			if (!may_change_field ('r')) {
				fprintf (stderr,
				         _("%s: Permission denied.\n"), Prog);
				exit (E_NOPERM);
			}
			rflg = true;
			strtcpy_a(roomno, optarg);
			break;
		case 'R': /* no-op, handled in process_root_flag () */
			flags->chroot = true;
			break;
		case 'u':
			usage (E_SUCCESS);
			/*@notreached@*/break;
		case 'w':
			if (!may_change_field ('w')) {
				fprintf (stderr,
				         _("%s: Permission denied.\n"), Prog);
				exit (E_NOPERM);
			}
			wflg = true;
			strtcpy_a(workph, optarg);
			break;
		default:
			usage (E_USAGE);
		}
	}
}

/*
 * check_perms - check if the caller is allowed to add a group
 *
 *	Non-root users are only allowed to change their gecos field.
 *	(see also may_change_field())
 *
 *	Non-root users must be authenticated.
 *
 *	It will not return if the user is not allowed.
 */
static void check_perms (const struct passwd *pw)
{
#ifdef USE_PAM
	pam_handle_t *pamh = NULL;
	int retval;
	struct passwd *pampw;
#endif

	/*
	 * Non-privileged users are only allowed to change the gecos field
	 * if the UID of the user matches the current real UID.
	 */
	if (!amroot && pw->pw_uid != getuid ()) {
		fprintf (stderr, _("%s: Permission denied.\n"), Prog);
		closelog ();
		exit (E_NOPERM);
	}
#ifdef WITH_SELINUX
	/*
	 * If the UID of the user does not match the current real UID,
	 * check if the change is allowed by SELinux policy.
	 */
	if ((pw->pw_uid != getuid ())
	    && (check_selinux_permit (Prog) != 0)) {
		fprintf (stderr, _("%s: Permission denied.\n"), Prog);
		closelog ();
		exit (E_NOPERM);
	}
#endif

#ifndef USE_PAM
	/*
	 * Non-privileged users are optionally authenticated (must enter the
	 * password of the user whose information is being changed) before
	 * any changes can be made. Idea from util-linux chfn/chsh.
	 * --marekm
	 */
	if (!amroot && getdef_bool ("CHFN_AUTH")) {
		passwd_check(pw->pw_name, pw->pw_passwd);
	}

#else				/* !USE_PAM */
	pampw = getpwuid (getuid ()); /* local, no need for xgetpwuid */
	if (NULL == pampw) {
		fprintf (stderr,
		         _("%s: Cannot determine your user name.\n"),
		         Prog);
		exit (E_NOPERM);
	}

	retval = pam_start (Prog, pampw->pw_name, &conv, &pamh);

	if (PAM_SUCCESS == retval) {
		retval = pam_authenticate (pamh, 0);
	}

	if (PAM_SUCCESS == retval) {
		retval = pam_acct_mgmt (pamh, 0);
	}

	if (PAM_SUCCESS != retval) {
		fprintf (stderr, _("%s: PAM: %s\n"),
		         Prog, pam_strerror (pamh, retval));
		SYSLOG((LOG_ERR, "%s", pam_strerror (pamh, retval)));
		if (NULL != pamh) {
			(void) pam_end (pamh, retval);
		}
		exit (E_NOPERM);
	}
	(void) pam_end (pamh, retval);
#endif				/* USE_PAM */
}

/*
 * update_gecos - update the gecos fields in the password database
 *
 *	Commit the user's entry after changing her gecos field.
 */
static void update_gecos(const char *user, char *gecos, const struct option_flags *flags)
{
	const struct passwd *pw;	/* The user's password file entry */
	struct passwd pwent;		/* modified password file entry */
	bool process_selinux;

	process_selinux = !flags->chroot;

	/*
	 * Before going any further, raise the ulimit to prevent colliding
	 * into a lowered ulimit, and set the real UID to root to protect
	 * against unexpected signals. Any keyboard signals are set to be
	 * ignored.
	 */
	if (setuid (0) != 0) {
		fputs (_("Cannot change ID to root.\n"), stderr);
		SYSLOG ((LOG_ERR, "can't setuid(0)"));
		fail_exit (E_NOPERM, process_selinux);
	}
	pwd_init ();

	/*
	 * The passwd entry is now ready to be committed back to the
	 * password file. Get a lock on the file and open it.
	 */
	if (pw_lock () == 0) {
		fprintf (stderr,
		         _("%s: cannot lock %s; try again later.\n"),
		         Prog, pw_dbname ());
		fail_exit (E_NOPERM, process_selinux);
	}
	pw_locked = true;
	if (pw_open (O_RDWR) == 0) {
		fprintf (stderr,
		         _("%s: cannot open %s\n"), Prog, pw_dbname ());
		fail_exit (E_NOPERM, process_selinux);
	}

	/*
	 * Get the entry to update using pw_locate() - we want the real one
	 * from /etc/passwd, not the one from getpwnam() which could contain
	 * the shadow password if (despite the warnings) someone enables
	 * AUTOSHADOW (or SHADOW_COMPAT in libc).  --marekm
	 */
	pw = pw_locate (user);
	if (NULL == pw) {
		fprintf (stderr,
		         _("%s: user '%s' does not exist in %s\n"),
		         Prog, user, pw_dbname ());
		fail_exit (E_NOPERM, process_selinux);
	}

	/*
	 * Make a copy of the entry, then change the gecos field. The other
	 * fields remain unchanged.
	 */
	pwent = *pw;
	pwent.pw_gecos = gecos;

	/*
	 * Update the passwd file entry. If there is a DBM file, update that
	 * entry as well.
	 */
	if (pw_update (&pwent) == 0) {
		fprintf (stderr,
		         _("%s: failed to prepare the new %s entry '%s'\n"),
		         Prog, pw_dbname (), pwent.pw_name);
		fail_exit (E_NOPERM, process_selinux);
	}

	/*
	 * Changes have all been made, so commit them and unlock the file.
	 */
	if (pw_close (process_selinux) == 0) {
		fprintf (stderr, _("%s: failure while writing changes to %s\n"), Prog, pw_dbname ());
		SYSLOG ((LOG_ERR, "failure while writing changes to %s", pw_dbname ()));
		fail_exit (E_NOPERM, process_selinux);
	}
	if (pw_unlock (process_selinux) == 0) {
		fprintf (stderr, _("%s: failed to unlock %s\n"), Prog, pw_dbname ());
		SYSLOG ((LOG_ERR, "failed to unlock %s", pw_dbname ()));
		/* continue */
	}
	pw_locked = false;
}

/*
 * get_old_fields - parse the old gecos and use the old value for the fields
 *                  which are not set on the command line
 */
static void get_old_fields (const char *gecos)
{
	char        *p;
	char        old_gecos[BUFSIZ];
	const char  *f;

	strtcpy_a(old_gecos, gecos);
	p = old_gecos;

	f = strsep(&p, ",");
	if (!fflg)
		strcpy(fullnm, f ?: "");

	f = strsep(&p, ",");
	if (!rflg)
		strcpy(roomno, f ?: "");

	f = strsep(&p, ",");
	if (!wflg)
		strcpy(workph, f ?: "");

	f = strsep(&p, ",");
	if (!hflg)
		strcpy(homeph, f ?: "");

	/* Anything left over is "slop".  */
	if (!oflg)
		strcpy(slop, p ?: "");
}

/*
 * check_fields - check all of the fields for valid information
 *
 *	It will not return if a field is not valid.
 */
static void check_fields (bool process_selinux)
{
	int err;
	err = valid_field (fullnm, ":,=\n");
	if (err > 0) {
		fprintf (stderr, _("%s: name with non-ASCII characters: '%s'\n"), Prog, fullnm);
	} else if (err < 0) {
		fprintf (stderr, _("%s: invalid name: '%s'\n"), Prog, fullnm);
		fail_exit (E_NOPERM, process_selinux);
	}
	err = valid_field (roomno, ":,=\n");
	if (err > 0) {
		fprintf (stderr, _("%s: room number with non-ASCII characters: '%s'\n"), Prog, roomno);
	} else if (err < 0) {
		fprintf (stderr, _("%s: invalid room number: '%s'\n"),
		         Prog, roomno);
		fail_exit (E_NOPERM, process_selinux);
	}
	if (valid_field (workph, ":,=\n") != 0) {
		fprintf (stderr, _("%s: invalid work phone: '%s'\n"),
		         Prog, workph);
		fail_exit (E_NOPERM, process_selinux);
	}
	if (valid_field (homeph, ":,=\n") != 0) {
		fprintf (stderr, _("%s: invalid home phone: '%s'\n"),
		         Prog, homeph);
		fail_exit (E_NOPERM, process_selinux);
	}
	err = valid_field (slop, ":\n");
	if (err > 0) {
		fprintf (stderr, _("%s: '%s' contains non-ASCII characters\n"), Prog, slop);
	} else if (err < 0) {
		fprintf (stderr,
		         _("%s: '%s' contains illegal characters\n"),
		         Prog, slop);
		fail_exit (E_NOPERM, process_selinux);
	}
}

/*
 * chfn - change a user's password file information
 *
 *	This command controls the GECOS field information in the password
 *	file entry.
 *
 *	The valid options are
 *
 *	-f	full name
 *	-r	room number
 *	-w	work phone number
 *	-h	home phone number
 *	-o	other information (*)
 *
 *	(*) requires root permission to execute.
 */
int main (int argc, char **argv)
{
	char                 new_gecos[80];
	char                 *user, *p, *e;
	const struct passwd  *pw;
	struct option_flags  flags = {.chroot = false};
	bool                 process_selinux;

	sanitize_env ();
	check_fds ();

	log_set_progname(Prog);
	log_set_logfd(stderr);

	if (getuid() == 0) {
		(void) setlocale (LC_ALL, "");
		(void) bindtextdomain (PACKAGE, LOCALEDIR);
		(void) textdomain (PACKAGE);
	}

	process_root_flag ("-R", argc, argv);

	/*
	 * This command behaves different for root and non-root
	 * users.
	 */
	amroot = (getuid () == 0);

	OPENLOG (Prog);

	/* parse the command line options */
	process_flags (argc, argv, &flags);
	process_selinux = !flags.chroot;

	/*
	 * Get the name of the user to check. It is either the command line
	 * name, or the name getlogin() returns.
	 */
	if (optind < argc) {
		if (!is_valid_user_name (argv[optind])) {
			fprintf (stderr, _("%s: Provided user name is not a valid name\n"), Prog);
			fail_exit (E_NOPERM, process_selinux);
		}
		user = argv[optind];
		pw = xgetpwnam (user);
		if (NULL == pw) {
			fprintf (stderr, _("%s: user '%s' does not exist\n"), Prog,
			         user);
			fail_exit (E_NOPERM, process_selinux);
		}
	} else {
		pw = get_my_pwent ();
		if (NULL == pw) {
			fprintf (stderr,
			         _("%s: Cannot determine your user name.\n"),
			         Prog);
			SYSLOG ((LOG_WARN, "Cannot determine the user name of the caller (UID %lu)",
			         (unsigned long) getuid ()));
			fail_exit (E_NOPERM, process_selinux);
		}
		user = xstrdup (pw->pw_name);
	}

	/* Check that the caller is allowed to change the gecos of the
	 * specified user */
	check_perms (pw);

	/* If some fields were not set on the command line, load the value from
	 * the old gecos fields. */
	get_old_fields (pw->pw_gecos);

	/*
	 * If none of the fields were changed from the command line, let the
	 * user interactively change them.
	 */
	if (!fflg && !rflg && !wflg && !hflg && !oflg) {
		printf (_("Changing the user information for %s\n"), user);
		new_fields ();
	}

	/*
	 * Check all of the fields for valid information
	 */
	check_fields (process_selinux);

	/* Build the new GECOS field by plastering all the pieces together.  */
	p = new_gecos;
	e = new_gecos + countof(new_gecos);
	p = stpeprintf(p, e, "%s", fullnm);
	p = stpeprintf(p, e, ",%s", roomno);
	p = stpeprintf(p, e, ",%s", workph);
	p = stpeprintf(p, e, ",%s", homeph);
	if (!streq(slop, ""))
		p = stpeprintf(p, e, ",%s", slop);

	if (p == e || p == NULL) {
		fprintf (stderr, _("%s: fields too long\n"), Prog);
		fail_exit (E_NOPERM, process_selinux);
	}

	/* Rewrite the user's gecos in the passwd file */
	update_gecos (user, new_gecos, &flags);

	SYSLOG ((LOG_INFO, "changed user '%s' information", user));

	nscd_flush_cache ("passwd");
	sssd_flush_cache (SSSD_DB_PASSWD);

	closelog ();
	exit (E_SUCCESS);
}

